Sensei Upsell: Update upgrade URL based on hosting provider

bogiii opened the pull request:
Resolves https://github.com/Automattic/sensei/issues/7784

## Proposed Changes

* Introduces a central place for both BE and FE to properly manage the upsell link
  * [includes/class-sensei-pro-upsell.php](https://github.com/Automattic/sensei/pull/7790/files#diff-539e6a7398a6f0ae5b71550330b749f7ba4f8849b3907ddf29bd552a18f9ea26)
  * [assets/admin/helpers.js](https://github.com/Automattic/sensei/pull/7790/files#diff-2d87186ff83194bbff5e6a40bbf4c0a1798a1902bf7a8357cc7ceaab9bdbe7ec)
* Update Sensei Pro upgrade URL to handle WPCOM-hosted sites correctly
* The logic recognizes if the hosting is WPCOM or other; based on it, builds the upsell URL

## Testing Instructions

- Checkout this branch and run the `npm run build` command
- Make sure you have a business plan site on your WPCOM account
- Upload and install the `sensei-lms.zip` file generated by the first step on Dotcom
- Create a new course
- Navigate through the Sensei plugin and check the upsell/upgrade button ("Get Sensei Pro")
- Check if you are redirected to the `https://wordpress.com/plugins/sensei-pro/{SITE_SLUG}`

Some of the affected CTAs:
<img width="852" alt="Screenshot 2025-03-31 at 17 12 00" src="https://github.com/user-attachments/assets/fa176702-4854-4c7d-bb19-26f828e503c0" />
<img width="1074" alt="Screenshot 2025-03-31 at 17 17 47" src="https://github.com/user-attachments/assets/ba8a8c4a-fcf9-4e39-b902-3fd868907c06" />
<img width="977" alt="Screenshot 2025-03-31 at 18 09 57" src="https://github.com/user-attachments/assets/68b74ef6-9c24-4a67-9e66-67d746f0371d" />

## Pre-Merge Checklist
<!-- Complete applicable items on this checklist **before** merging. Items that are not applicable can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed. -->
- [x] PR title and description contain sufficient detail and accurately describe the changes
- [ ] Adheres to coding standards ([PHP](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/), [JavaScript](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/), [CSS](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/css/), [HTML](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/html/))
- [ ] All strings are translatable (without concatenation, handles plurals)
- [ ] Follows our naming conventions (P6rkRX-4oA-p2)
- [ ] Hooks (p6rkRX-1uS-p2) and functions are documented
- [ ] New UIs are responsive and use a [mobile-first approach](https://zellwk.com/blog/how-to-write-mobile-first-css/)
- [ ] Code is tested on the minimum supported PHP and WordPress versions

bogiii committed:
Create helper methods

bogiii committed:
Update upgrade URL based on hosting provider

bogiii committed:
Changelog

bogiii committed:
Improve isWpcom function logic

bogiii committed:
Update isWpcom method; remove wrong statement

Imran92 commented on assets/admin/editor-wizard/helpers.js:
```diff
@@ -188,3 +189,33 @@ export const useHideEditorWizardUpsell = () => {
+ */
+export const isWpcom = () => {
+	// Check for Jetpack connection
+	if ( window?.JetpackScriptData?.connection?.connectionStatus?.isActive ) {
+		return true;
+	}
```
This logic will cause `isWpcom` to return true on self-hosted Jetpack sites too I think. Which will probably cause the WPCOM url to open instead of the Sensei-LMS one. That's probably not expected, am I getting it right?

bogiii replied:
You are right. I overlooked it...
I'll try another approach. Thanks!

bogiii replied:
Hey @Imran92,

I have updated the logic and found a better field to check, which is only presented in the WPCOM env. Please re-review when you find some time.

bogiii committed:
Improve isWpcom method

renatho commented on assets/admin/editor-wizard/helpers.js:
```diff
@@ -188,3 +189,30 @@ export const useHideEditorWizardUpsell = () => {
+ *
+ * @return {string} Site slug.
+ */
+export const useSiteSlug = () => {
```
I don't think the `editor-wizard` would be the best place to add this since we should use it in some different parts.

Maybe it was a confusion because of the screenshot I added to the issue. That was supposed to be only an example, and we have other upsells in the plugin that I didn't map as part of the issue. Sorry! 😜

So maybe we could add these functions in some place like `assets/admin`?

renatho replied:
I think the best way to find all the places with upsells is through a search in all the codebase for `https://senseilms.com/sensei-pro`.

bogiii replied:
> Maybe it was a confusion because of the screenshot I added to the issue. That was supposed to be only an example, and we have other upsells in the plugin that I didn't map as part of the issue. Sorry! 😜

No problem. I'm going to update the logic and cover all upsell cases. 👌

renatho commented on assets/admin/editor-wizard/helpers.js:
```diff
@@ -188,3 +189,30 @@ export const useHideEditorWizardUpsell = () => {
+ *
+ * @return {string} Site slug.
+ */
+export const useSiteSlug = () => {
```
Not sure if `SiteSlug` makes sense for other scenarios out of WPCOM. Maybe to make it more clear, could we call it as `useWpcomSiteSlug`?

We could also use only this hook in the components, returning `null` if it's not a WPCOM site. WDYT? Just to simplify since we'll need to have this logic of slug + wpcom check in many places.

renatho commented on assets/admin/editor-wizard/steps/course-upgrade-step.js:
```diff
@@ -98,10 +99,34 @@ const FeatureItem = ( { children } ) => (
+ * @return {string} The upgrade URL.
+ */
+const getUpgradeUrl = ( siteSlug, isWpcomHosting ) => {
+	const sensieParams = new URLSearchParams( {
```
```suggestion
	const senseiParams = new URLSearchParams( {
```

And we also need to rename where we're using the variable.

bogiii committed:
Create Sensei_Pro_Upsell class

bogiii committed:
Provide upsellUrl to sensei_home const

bogiii committed:
Make upsellUrl constant avaiable globally through the Sensei plugin

bogiii committed:
Refactor upsell URL handling to be globally accessible

- Move upsell URL to global admin data via window.sensei_admin
- Create reusable getSenseiProUpsellUrl helper function
- Update course upgrade step to use global helper

bogiii committed:
Replace hardcoded upsell links on the JS side

bogiii committed:
Use centralized upsell URL generation within PHP files

Replace hardcoded URLs with calls to Sensei_Pro_Upsell::get_base_url() and
get_default_utm_params() across the codebase for better maintainability and
consistency. Updated files:

- class-sensei-home-task-pro-upsell.php
- class-sensei-course.php
- class-sensei-groups-landing-page.php
- class-sensei-lesson.php
- class-sensei-settings-api.php
- class-sensei-course-theme-template-selection.php
- class-email-list-table.php

Also added proper use statements for Sensei_Pro_Upsell and add_query_arg
functions where needed.

bogiii committed:
Fix PHP linting issues

bogiii committed:
Update/fix failing test

bogiii commented:
Hey @renatho, I've updated my approach and created a 'single source of truth' related to the upsell URL generation. It is consumed on both the FE and BE sides.

When you find some time, please take a look. Thanks!

renatho commented on assets/admin/helpers.js:
```diff
@@ -0,0 +1,18 @@
+export const getSenseiProUpsellUrl = ( campaign = '' ) => {
+	const { upsellUrl } = window.sensei_admin;
+
+	const sensieParams = new URLSearchParams( {
```
```suggestion
	const senseiParams = new URLSearchParams( {
```

renatho commented on assets/admin/helpers.js:
```diff
@@ -0,0 +1,18 @@
+		utm_campaign: campaign,
+	} );
+
+	return `${ upsellUrl }?${ sensieParams.toString() }`;
```
```suggestion
	return `${ upsellUrl }?${ senseiParams.toString() }`;
```

renatho commented on includes/class-sensei-settings-api.php:
```diff
@@ -217,7 +217,10 @@ private function render_content_drip_settings() {
 		$header             = __( 'Get Sensei Pro', 'sensei-lms' );
 		$text               = __( 'Keep students engaged and improve knowledge retention by setting a delivery schedule for course content.', 'sensei-lms' );
-		$url                = 'https://senseilms.com/sensei-pro/?utm_source=plugin_sensei&utm_medium=upsell&utm_campaign=settings_content_drip';
+		$url                = add_query_arg(
```
Since we'll also need to compose the URL with `get_default_utm_params` and `get_base_url`, would it be simpler if we had a method similar to the `getSenseiProUpsellUrl` that you created in JS?

renatho commented on includes/class-sensei-pro-upsell.php:
```diff
@@ -0,0 +1,48 @@
+ * File containing the Sensei_Pro_Upsell class.
+ *
+ * @package sensei-lms
+ * @since   4.23.0
```
In Sensei, when writing new code and marking a version (`since`, `deprecated`, etc.), you can use the `$$next-version$$`. As part of the release process, we'll update all these references with the next version. :)

bogiii replied:
The version number has been fixed.

m1r0 replied:
I see the version is still `4.23.0`. Probably you forgot to change it to the `$$next-version$$` placeholder?

bogiii replied:
Replaced

renatho commented on assets/admin/helpers.js:
```diff
@@ -0,0 +1,18 @@
+		utm_campaign: campaign,
+	} );
+
+	return `${ upsellUrl }?${ sensieParams.toString() }`;
```
While reading the logic, I realized something.

The tracking params won't be useful when pointing to the wordpress.com URL, and we would lose this tracking. Unless we also register it on wordpress.com, and I'm not aware.

A solution I was thinking of would be a URL on senseilms.com that we catch the user, track, and redirect to wordpress.com. Maybe it could be a new WordPress page with a script to get the params and redirect.

WDYT? 🤔

bogiii replied:
@m1r0, what do you think about this?

m1r0 replied:
As far as I know, wordpress.com is using UTM parameters for tracking ( PCYsg-lOL-p2 ), which means we should be able to track the upsells. Because this is a somewhat important metric, it's probably worth asking someone from marketing.

@dzver, I'm not sure if this is in your domain, but do you happen to know if we can track links similar to `wordpress.com/plugins/sensei-pro/?utm_source=plugin_sensei&utm_medium=upsell&utm_campaign=course_editor_wizard`? For context, previously, all Sensei Pro upsell links were pointing to `senseilms.com` + UTMs. With this PR, we're detecting if the user is coming from wordpress.com and upsell links are targeting `wordpress.com` + the same UTMs. Do you think that's fine?

dzver replied:
Are you trying to log events for senseilms.com from wordpress.com/plugins? Gtag should support using two different IDs. However, I'm  unfamiliar with this particular page and if it uses the utm parameters for anything.

m1r0 replied:
> Are you trying to log events for senseilms.com from wordpress.com/plugins?

We're trying to figure out the best approach for tracking all the upsell links in the Sensei LMS plugin. The issue is that, with this PR, the upsell link is changing based on whether the user is on wp.com or not. To make it more clear:
* Upsell links without this PR go to: `senseilms.com/sensei-pro/?utm_source=plugin_sensei`
* Upsell links with this PR go to:
  * For wp.com users: `wordpress.com/plugins/sensei-pro/?utm_source=plugin_sensei`
  * For self hosted users: `senseilms.com/sensei-pro/?utm_source=plugin_sensei`

I guess the question is, what do you think would be the best approach for tracking here? Do you think [this](https://github.com/Automattic/sensei/pull/7790#discussion_r2025015156) is a good approach? If you're unsure, do you happen to know who we to ask?

m1r0 replied:
FYI, if it makes sense, we have the option to switch to Tracks (pixel.wp.com) as well.

dzver replied:
I think the utm parameters will just work, however the IDs are different and I'm not sure if that's fine. Tagging @debbiengWP 

Some questions from me (Debbie might have better questions here)

- What do we want to track
- Where are we tracking it? Is it Google Analytics?
- Are the IDs preserved?

https://senseilms.com/sensei-pro/ uses G-MV34CC1ZMY 

https://wordpress.com/plugins/sensei-pro uses G-1H4VG5F5JF (dotcom)

m1r0 replied:
Oh, right, I forgot about these IDs. Thanks for bringing that up! I have zero context for how those work, so any help would be great. 🙏

dzver replied:
The G-something is a Measurement ID tied to a GA4 property. Once the domain changes, the SSR landing page plugins/sensei-pro will switch to using the wpcom property.

Yes, you can send data to multiple GA4 Measurement IDs from the same request as I mentioned earlier in the thread but it will require some coding.

m1r0 replied:
@burtrw Do you happen to know where we're tracking this and if implementing the above makes sense?

burtrw replied:
My memory is fuzzy, but if you keep the UTMs the same, reports like this should still work:

https://analytics.google.com/analytics/web/?authuser=1#/p314881486/reports/explorer?params=_u..nav%3Dmaui%26_r.explorerCard..selmet%3D%5B%22sessions%22%5D%26_r.explorerCard..seldim%3D%5B%22sessionManualSource%22%5D&collectionId=3896115603&r=manual

bogiii replied:
Thanks @burtrw 🙌 .

@m1r0, then this solution looks right.

m1r0 replied:
Thanks, Ronnie! You're my hero! 🫶

Let's continue with this approach, and if needed, we'll make the ID changes in another iteration.

renatho commented on assets/blocks/course-outline/outline-block/outline-edit.js:
```diff
@@ -27,8 +27,7 @@ import ExistingLessonsModal from './existing-lessons-modal';
 import useSenseiProSettings from './use-sensei-pro-settings';
 import { applyFilters } from '@wordpress/hooks';
-
-const SENSEI_PRO_LINK = 'https://senseilms.com/sensei-pro/';
```
Very nice to have this unified solution you created, so we make sure we track all the upsell links, that we had missed here, for example! =)

renatho commented on includes/class-sensei-course.php:
```diff
@@ -265,9 +265,18 @@ public function showcase_courses_screen() {
+					<?php
+					$upsell_args = array_merge(
+						Sensei_Pro_Upsell::get_default_utm_params(),
+						[ 'utm_campaign' => 'showcase' ]
```
Shouldn't it be in the `get_default_utm_params` argument? And the same for the other cases where we have this.

renatho reviewed (commented):
Thanks for working on this, @bogiii!
The solution looks great! I just realized a few things and added in comments in the code.

bogiii committed:
Fix the plugin version

bogiii committed:
Fix typo

bogiii committed:
Create an upsell url compoposer method

bogiii committed:
Replace URL compose with the Sensei_Pro_Upsell::get_sensei_pro_upsell_url

m1r0 commented on includes/class-sensei-admin.php:
```diff
@@ -444,13 +444,32 @@ public function register_scripts( $hook ) {
+	 *
+	 * This method adds global data to the admin scripts, such as the upsell URL.
+	 *
+	 * @since 4.24.0
+	 * @access private
```
These can be removed because it's a private method.

As standard, we use `@access`/`@internal` and `@since` on public methods. Some more info can be found here: p6rkRX-4Og-p2.

Specifically for `@since`, we use a special tag to indicate the next released version, `$$next-version$$`, which is replaced during the release process. More here: p6rkRX-67G-p2

bogiii replied:
Thanks for the feedback. I've removed it. 👍

m1r0 commented on includes/class-sensei-pro-upsell.php:
```diff
@@ -0,0 +1,63 @@
+ *
+ * Handles the generation of Pro upsell URLs.
+ *
+ * @since 4.24.6
```
Same here, the version should be `$$next-version$$`.

bogiii replied:
I got the same feedback from Renatho, and I didn't know about the special tag, so I manually entered the next version. :)

It's replaced now. Thanks

m1r0 replied:
Haha, yeah, it's a bit confusing practice. :)

Thanks for the updates!

m1r0 commented on includes/internal/emails/class-email-list-table.php:
```diff
@@ -8,6 +8,8 @@
 
 use Sensei_List_Table;
+use Sensei_Pro_Upsell;
+use function add_query_arg;
```
Is the `use function add_query_arg` needed? I'm guessing it should continue working without it.

bogiii replied:
Removed

m1r0 commented on sensei-lms.php:
```diff
@@ -86,7 +86,8 @@ function is_sensei_activating( $activating_plugin, $plugin ) {
 
-require_once dirname( __FILE__ ) . '/includes/class-sensei-dependency-checker.php';
+require_once __DIR__ . '/includes/class-sensei-dependency-checker.php';
+require_once __DIR__ . '/includes/class-sensei-pro-upsell.php';
```
There is an effort to move new internal classes (internal because they are not meant for use by other developers) to the `includes/internal` folder. Probably the `Sensei\Internal\Services` namespace would be a good place for this. Can you please move it there and add the `@internal` tag to the class? 🙏 

The `require_once` won't be needed afterward because the composer autoloader will take care of it.

bogiii replied:
Done following your suggestion. Thanks. 🙌

m1r0 reviewed (commented):
Overall it looks great! Thanks for working on this and extracting the logic in a well defined structure!

I have a few very minor suggestions.

Also, I've noticed the Home screen upsell links are still pointing to `senseilms.com`. Is this the case on your end as well?

![image](https://github.com/user-attachments/assets/bfbd373f-cf14-4778-9a56-ffb1fe16a694)

bogiii committed:
Remove since and access tags from the localize_admin_data function

bogiii committed:
Replace version with the next-version tag

bogiii committed:
Remove unused function

bogiii committed:
Move Sensei_Pro_Upsell class to the internal services folder

bogiii committed:
Introduce getCheckoutUrl based on the hosting provider

bogiii commented:
> Also, I've noticed the Home screen upsell links are still pointing to senseilms.com. Is this the case on your end as well?

@m1r0 You were right. I missed it because of the checkout link. I've introduced another method for getting the checkout URL based on the hosting provider type.

bogiii committed:
Remove testing if statement

bogiii committed:
Use proper namespace

bogiii committed:
Use proper namespace

m1r0 commented on includes/admin/home/tasks/task/class-sensei-home-task-pro-upsell.php:
```diff
@@ -74,7 +76,12 @@ public static function mark_completed_and_redirect() {
 
 		// phpcs:ignore WordPress.Security.SafeRedirect.wp_redirect_wp_redirect -- We're redirecting to an external URL.
-		wp_redirect( 'https://senseilms.com/sensei-pro/?utm_source=plugin_sensei&utm_medium=upsell&utm_campaign=sensei-home' );
+		wp_redirect(
```
A [standard practice](https://developer.wordpress.org/reference/functions/wp_redirect/) is to have an `exit` after a `wp_redirect`. It prevents unexpected side effects. Let me know if there's a reason not to have it.

m1r0 replied:
Now I see that this logic is not new, sorry. FYI, I've added the exit.

m1r0 committed:
Add missing `exit` after `wp_redirect`

m1r0 commented:
LTMG! Thanks for the updates, Bogi!

I'm merging it to start testing on WP 6.8.
